### PR TITLE
Add ability to generate sprintly zip

### PR DESCRIPTION
**Pull request by stephenmichaelf**
As part of task lock down, we want to release a zip of all the tasks for a release.

This will be generated and uploaded for n-1 release(after a release has fully rolled out).

### CODE REVIEW

**Comment by damccorm** [comment]
How important is it that this match exactly what gets published? I'm a little worried that even if we do this at the same time as our courtesy push, people might push package updates after that.

File: make.js [hunk 1]
```diff
@@ -577,3 +578,112 @@ target.bump = function() {
...
+    fs.mkdirSync(tempWorkspaceDirectory);
+
+    console.log('Writing packages.config file');
+
+    var packagesConfigPath = `${tempWorkspaceDirectory}\\packages.config`;
+    var packagesConfigContent = '<?xml version="1.0" encoding="utf-8"?>\n';
+    packagesConfigContent += '<packages>\n';
+
+    taskStringArr.forEach(function (taskString) {
+        packagesConfigContent += taskString;
```
Comment (damccorm) [inline make.js]: Should these be separated by `\n`?

File: make.js [hunk 2]
```diff
@@ -577,3 +578,112 @@ target.bump = function() {
         }
     });
 }
+
+// Generate spritly zip
```
Comment (damccorm) [inline make.js]: Nit: `sprintly`

**Comment by stephenmichaelf** [comment]
> How important is it that this match exactly what gets published? I'm a little worried that even if we do this at the same time as our courtesy push, people might push package updates after that.

It doesn't have to be exactly the same. We are just going for "this is approximately what was released for this sprint".

Also the plan is to not generate until a sprint has fully rolled out so we should be OK with people not making changes after. It should be very unlikely.

**Comment by stephenmichaelf** [comment]
> Generally looks good - would be awesome if we could eventually do this in scheduled CI

That would be great, the one blocker there is permissions. This needs permissions to restore from packaging.

File: make.js (hunk 1 above)
Comment (stephenmichaelf) [reply make.js]: The line already contains newline character.

Comment (damccorm) [reply make.js]: I'm a little confused by this. Isn't the basic flow (with reduced logic):
```
var dependenciesArr = dependencies.split('\n');
var taskStringArr = [];
dependenciesArr.forEach(function (currentDep) {
    taskStringArr.push(currentDep);
}
taskStringArr.forEach(function (taskString) {
    packagesConfigContent += taskString;
}
```

Wouldn't the `\n` characters get removed when we split on it?

**Comment by damccorm** [comment]
> Also the plan is to not generate until a sprint has fully rolled out so we should be OK with people not making changes after. It should be very unlikely.

Hm, ok - so this would be done separately from the courtesy push then? I don't love adding another manual step, that doubly makes me want to automate this if possible. That shouldn't block us merging here though.

> This needs permissions to restore from packaging.

Can't we just add a PAT as a secret to do that? We could add a new Pipeline to do that which doesn't run CI/PR it only runs scheduled builds. That way we don't need to worry about security risks at all really.

I'm also pretty sure our CI builds currently have PAT access to packaging (not 100% certain - don't quote me on that though)

**Comment by damccorm** [comment]
Note - CI changes don't have to be (and probably shouldn't be) in scope for this PR - just want to get some semblance of a plan before moving forward

**Comment by stephenmichaelf** [comment]
> Hm, ok - so this would be done separately from the courtesy push then? I don't love adding another manual step, that doubly makes me want to automate this if possible. That shouldn't block us merging here though.
> 
> >

Yes. Courtesy push is before VSO branches for the next sprint. Then people make changes to that branch throughout the entire next sprint while it's rolling out. This will be once that sprint is fully rolled out, then we generate this zip (at least that's the plan for now, that is when the tasks should be the most stable).

> Can't we just add a PAT as a secret to do that? We could add a new Pipeline to do that which doesn't run CI/PR it only runs scheduled builds. That way we don't need to worry about security risks at all really.

Potentially. I will add a follow up ticket to see how we can make this as easy/consistent as possible. I think a PAT will work but it would have to be on a private project.

File: make.js (hunk 1 above)
Comment (stephenmichaelf) [reply make.js]: Oh yeah it's possible, the file is intermittent (I use it to restore packages then delete it) so we should be OK. I will take a look.

**Pull request merged**